haystack/preview/components/samplers/top_p.py

```diff
@@ -114,41 +117,66 @@ def run(self, query: str, documents: List[Document], top_p: Optional[float] = No
         if not documents:
             return []
 
-        if self.cross_encoder is None:
+        if self.model_name_or_path and self.model is None:
             raise ComponentError(
                 f"The component {self.__class__.__name__} not warmed up. Run 'warm_up()' before calling 'run()'."
             )
+        if not self.model_name_or_path and not self.has_scores(documents):
+            raise ComponentError(
+                f"The component {self.__class__.__name__} requires scores in the documents' metadata. "
+                f"Either set 'model_name_or_path' to scored documents or add scores to the documents."
+            )
 
-        # prepare the data for the cross encoder
-        query_doc_pairs = [[query, doc.text] for doc in documents]
+        needs_scoring = self.model_name_or_path and not self.has_scores(documents)
+        if needs_scoring:
+            # prepare the data for the model
+            query_doc_pairs = [[query, doc.text] for doc in documents]
+            features = self.tokenizer(query_doc_pairs, padding=True, truncation=True, return_tensors="pt")
 
-        # compute the similarity scores for these combinations
-        similarity_scores = self.cross_encoder.predict(query_doc_pairs)
+            with torch.no_grad():
```

Comment:
I would make this `with torch.inference_mode()` since that is the preferred option over `torch.no_grad()` I believe.